main.go

```diff
@@ -7,142 +7,150 @@ import (
+		}
 
-		fail_a := make(chan struct{})
-		fail_b := make(chan struct{})
-		fail := make(chan struct{})
+		// Allocate a cancellation channel for each target's HTTP request.
+		cancels := make([]chan struct{}, len(targets))
 
-		cancel_a := make(chan struct{})
-		cancel_b := make(chan struct{})
+		// Allocate a failure channel for each target's HTTP request.
+		fails := make([]chan struct{}, len(targets))
 
 		r.RequestURI = ""
 		r.URL.Scheme = "http"
 
-		go func() {
-			req_a := *r
-			req_a.URL.Host = ua.Host
-			req_a.URL, _ = url.Parse(r.URL.String())
-			req_a.Cancel = cancel_a
+		// Create and send out an HTTP request to each target.
+		for i := range targets {
+			req := *r
+			req.URL.Host = urls[i].Host
+			req.URL, _ = url.Parse(r.URL.String())
+			cancels[i] = make(chan struct{})
+			fails[i] = make(chan struct{})
+			req.Cancel = cancels[i]
+			fail := fails[i]
+
+			channel := channels[i]
+
+			go func() {
+				rt := &http.Transport{DisableKeepAlives: true}
+				res, err := rt.RoundTrip(&req)
+				if err != nil {
+					log.Printf("request failed: %s\n", err)
+					close(fail)
+				} else if res.StatusCode >= 500 || res.StatusCode == 408 {
+					log.Printf("target (%s) unsatisfying status: %d", req.URL, res.StatusCode)
+					close(fail)
+				} else {
+					log.Printf("target (%s) responded with %d", req.URL, res.StatusCode)
+					channel <- res
+				}
+			}()
+		}
 
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_a)
-			if err != nil {
-				log.Printf("target A failed: %s", err)
-				close(fail_a)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target A unsatisfying status: %d", resp.StatusCode)
-				close(fail_a)
-			} else {
-				resp_a <- resp
+		// Listen to all requests' failure channels; issue a signal when all requests fail.
+		failed := make(chan struct{})
+		go func() {
+			for i := range fails {
+				<-fails[i]
 			}
+			close(failed)
 		}()
 
-		go func() {
-			req_b := *r
-			req_b.URL, _ = url.Parse(r.URL.String())
-			req_b.URL.Host = ub.Host
-			req_b.Cancel = cancel_b
+		// Listen to the request channels for a response.
+		cases := make([]reflect.SelectCase, len(channels)+1)
+		for i, ch := range channels {
+			cases[i] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ch)}
+		}
+		// Plus an extra channel for the failure case.
+		cases[len(channels)] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(failed)}
 
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_b)
-			if err != nil {
-				log.Printf("target B failed: %s", err)
-				close(fail_b)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target B unsatisfying status: %d", resp.StatusCode)
-				close(fail_b)
-			} else {
-				resp_b <- resp
+		// Wait for a successful response (or failures across the board).
+		for {
+			chosen, value, _ := reflect.Select(cases)
```

Comment:
i really feel like you could get away without using reflection here. Have one channel that a goroutine can send its index on when its done or something, then can grab  the response from an array by its index.